### ISSUE TITLE
The `history` var is extremely expensive and should be replaced by a `history item` command

### ISSUE THREAD

**Issue by krader1961** [krader1961/fish#17]
See also https://github.com/fish-shell/fish-shell/issues/4366

When I fixed issue #4353 it got me thinking about the cost of that var. On my main computer my history has 29,579 entries. Typing `set -g` takes a minimum of 170 ms (best of five runs). If I build fish with a one line change to exclude `history` from that output the run time drops to 2 ms. The following two functions makes the difference even more stark:

```
function histcopy
    set -l histcopy $history
    for i in (seq 1000)
        echo $i: $histcopy[$i]
    end
end
function hist
    for i in (seq 1000)
        echo $i: $history[$i]
    end
end
```

Then I ran this sequence several times:

```
histcopy
echo $CMD_DURATION
hist
echo $CMD_DURATION
```

Iterating over the copy takes 30 seconds and iterating over `$history` takes 190 seconds. Which means that ignoring all the common overhead expanding `$history[$i]` takes ~160 ms longer than expanding `$histcopy[$i]`.

A third way to show the overhead is this:

```
set -g histcopy $history
set --show histcopy
echo $CMD_DURATION
set --show history
echo $CMD_DURATION
```

The `history` var also does not expose when the job was run or anything else we might save about the job, now or in the future. A better interface would be modeled on `functions --details $funcname`. I propose `history item $i`. Or possibly `history entry $i` though I prefer `item`. This would allow us to expose everthing we currently save with each history entry and can be safely extended in the future if we save more data about the job (e.g., the CWD when the job was run).

Because of the backward incompatible nature of this change, and the trivialness of implementing it, this is a good candidate for fish 3.0.

**Comment by krader1961** [comment]
Note that fish 2.x is approximately 30 ms slower (20%) at expanding $history because it first constructs the history as a flat string then converts it to a vector. The improvement I did for fish 3.0 to always store fish vars as vectors internally avoids that extra representation conversion step.

**Comment by krader1961** [comment]
The `history count` command I just implemented does the same thing as the above functions in 1719 ms using this function and redirecting its output to /dev/null:

```
function hist2
    for i in (seq 1000)
        echo $i: (builtin history item $i)[1]
    end
end
```

That's 94.3% faster than the `histcopy` function and 99.1% faster than the `hist` function above.

**Comment by zx8** [comment]
Any chance we can get this merged upstream?

Happy to raise it as a PR on your behalf if you're busy! 😄

**Comment by krader1961** [comment]
> Any chance we can get this merged upstream?

Waiting for https://github.com/fish-shell/fish-shell/pull/4404 to be merged.

**Comment by krader1961** [comment]
Also, with floam busily adding negative value to the project by reopening issues that won't be implemented just because he reopens the issue I'm starting to think it's time to move on to a different shell with a smarter team managing its development. When there are really important improvements to be made in areas like this issue, not to mention job control management, how redirections are handled, etc., the fact the rest of the fish team thinks that renaming uppercase var names like FISH_HISTORY to lowercase is important leaves me shaking my head in disbelief.

**Comment by krader1961** [comment]
@zx8 and @alphapapa: In light of what floam has been doing recently (reopening every issue I closed because no one was ever going to implement the feature, such as https://github.com/fish-shell/fish-shell/issues/2111) I think I'm going to stop working on fish. I just can't deal with that level of stupid. More to the point, if the rest of the fish team thinks such issues are worth working on rather than fixing items like fish's job control and I/O redirection there is no hope for the project.

For fuck sake, they had been discussing a better way to handle option parsing for four years (https://github.com/fish-shell/fish-shell/issues/478) without actually implementing even a proof of concept implementation. After another six months went by and the deadline ridiculousfish had committed to for implementing the docopt solution came and went  I spent a few days and implemented`argparse` (see https://github.com/fish-shell/fish-shell/issues/4190).

**Comment by krader1961** [comment]
Sorry, @zx8 and @alphapapa, but just look at my recent exchange on [Gitter](https://gitter.im/fish-shell/fish-shell?at=59cc81e8177fb9fe7e10612a) with floam where I suggested he reopen issue https://github.com/fish-shell/fish-shell/issues/478. Despite that issue being almost five years old without a usable implementation. Despite the fact we now have a perfectly usable solution to parsing arguments in both builtin commands and fish script functions he reopened that issue. I'm definitely not going to contribute another change to the fish-shell project unless ridiculousfish tells floam he is not helping improve the project and revokes his commit privileges. Floam completely missed the sarcasm in my comment suggesting the fish community continue working on a proposal that hadn't been implemented in the nearly five years after it was opened and a perfectly usable solution had already been merged.

**Comment by krader1961** [comment]
@zx8 or @alphapapa:

You are free to take the commit mentioned earlier in this issue and turn it into a PR for the fish-shell project. I've decided that even if the other history change I was waiting to be merged ever gets merged I won't be submitting this, or any other, changes. I've given up on Fish and am looking for another shell. I'm tired of pointless arguments about whether `FISH_VERSION` should be renamed to `fish_version` or just `version`. Or whether completion enhancement requests that are more than a year old and will never be implemented (at least as a consequence of an open issue) should remain open or be closed. Basically, I don't think ridiculousfish shares my values for a project of this nature. So it is best that we part ways.

Sadly, I think you'll see fish revert to the state I found it in October 2015. I would actively maintain my own fork but I've decided that I am not willing to fix the serious problems with the code. Specifically with respect to how file descriptor redirections are handled including pipelines and command substitutions. Not to mention job control. The code for managing processes is awful and has numerous bugs and incorrect behavior. And those problems have been present for more than a decade. So the fact that core devs are making pointless changes like renaming `FISH_HISTORY` to `fish_history` leaves me wondering where the adults are.

**Comment by alphapapa** [comment]
Kurtis,

I'm sorry to hear that you've moving on from Fish.  I wish you and Aaron could get along well, but lacking that, I wish you could ignore your disagreements and work in parallel, if not together.  You've contributed a lot to the project, and your future contributions would be valuable.

I think there's room to disagree on whether to keep certain issues open or closed.  Personally, I lean toward the way Debian does it, leaving issues open as long as they are relevant, even if no one has touched them for a long time and no one has committed to working on them.  If nothing else, when someone else wants to report the same issue, it gives them a place to do so and a way to gauge interest.  To some people it seems messy to leave such issues open, but they can always be marked help-wanted and ignored until someone shows interest in working on them.  We can filter them--we have the technology.  :)

> I'm tired of pointless arguments about whether FISH_VERSION should be renamed to fish_version or just version. Or whether completion enhancement requests that are more than a year old and will never be implemented (at least as a consequence of an open issue) should remain open or be closed. Basically, I don't think ridiculousfish shares my values for a project of this nature. So it is best that we part ways.

You might be right that you have different values for the project, so that might be for the best.  On the other hand, there's nothing stopping work on both major and trivial issues at the same time, and people can always ignore issues they aren't interested in.  You could keep working on what you're interested in and let others fiddle with the bug tracker.

So, maybe I'm wrong, but I get the impression that this is more an issue of personality clashes than technical disagreements.  And that is disappointing, because you all have made a good team for a while now.  If there is some way that you all could agree to disagree on certain issues and let each other have their way where it isn't important to do it a certain way, it would be great to see you continue to be part of the team.

**Comment by floam** [comment]
> Floam completely missed the sarcasm in my comment suggesting the fish community continue working on a proposal that hadn't been implemented in the nearly five years after it was open and a perfectly usable solution had already been merged.

I did wrongly assume you were arguing in good faith there. I do wish you could be able to live and let live, but I guess that was not in the cards.

**Comment by krader1961** [comment]
> You could keep working on what you're interested in and let others fiddle with the bug tracker.

@zx8, Sadly, that isn't viable. That there are so many pointless, bike-shedding, arguments about such things as whether `FISH_VERSION` should be renamed to `fish_version` or `version` it tells me the project is hopeless given its current leadership. If there was actually a coherent, good, argument for renaming such vars I would not object to doing so. But that is not the case. As evidenced by the recent commit to rename `FISH_HISTORY` to `fish_history` despite the fact that out of all the uppercase var names (e.g., `CMD_DURATION`) that is possibly the only one that should rename all uppercase.

Keep in mind that when I found the project two years ago there were several pull-requests more than a year old. I closed or merged them. So being responsive to proposed changes is important to me. People like @floam seem to think that issues open for several years are as important as new issues. I disagree.

@floam, That you reopened https://github.com/fish-shell/fish-shell/issues/478 tells me you are very inexperienced and have lousy software engineering judgement. It doesn't matter that you thought I was arguing in good faith the issue should be reopened. I can deal with that since I do so regularly with respect to interns and other inexperienced, young, new hires. What I can't accept are the other Fish developers, especially ridiculousfish, letting your bad decisions stand unchallenged.

Note that the person who opened the [DocOpt issue](https://github.com/fish-shell/fish-shell/issues/478) subsequently started the [Elvish](https://github.com/elves/elvish/) project. Apparently even they gave up on on the Fish project. Which is what I'm doing.

**Comment by krader1961** [comment]
I'll be watching to see when fish 2.7.0 gets released since it should have already have happened. I'm not holding my breath for fish 3.0.0 occurring in the first half of 2018. Frankly, I'll be surprised if there is a fish 3.0.0 release in 2018 that contains much more than the changes already merged to the master branch.

**Comment by zx8** [comment]
Kurtis,

It's really hard to accept that you've chosen to leave the project, but I completely understand your rationale. In your position, I'd almost certainly do the same (although I'd probably have tried reaching out to ridiculousfish directly and expressing my concerns/proposing solutions first). 

As the majority of innovation and new/useful features this year have been contributed by you, I can't help but agree that fish's future doesn't look too bright without you on board. I'd also be very surprised if there are any other significant changes before 3.0 - assuming it's ever formally released, that is. 

I'll be keeping an eye on this fork + any future projects you move on to with the hope of finding another shell exclusively governed by you, or cross my fingers and hope ridiculousfish sees the light. 🙂

Thanks for all your hard work over the years, we appreciate it.

**Comment by krader1961** [comment]
@zx8, You are welcome to take this [commit]( https://github.com/krader1961/fish/commit/567e728dcba286ae0bf6df19e8412805c9d3ecc9) and try to get it merged into the fish-shell project. Even without attributing it to me. It would be a shame for it to be lost. But even if my [PR #4404](https://github.com/fish-shell/fish-shell/pull/4404) ever gets merged I'm not going to make the effort to get this enhancement merged. At least not without floam being kicked out of the dev team.

P.S., I did reach out to ridiculousfish regarding my concerns about floam. Just a few weeks after they were granted commit privileges. That was over a year ago. After that I had several "WTF?" disagreements with floam. Frankly, it was only the fact floam disappeared for a year that I continued to contribute. Now that he is again active and no one else is telling him he is adding negative value pushed me over the edge. But only because floam is distracting from  major problems.

**Comment by alphapapa** [comment]
Kurtis,

I'm really trying to remain neutral here, but I can't help but notice that you're insisting that floam be removed from the project, while he is not insisting the same about you.  And although you have made many positive contributions, there may be some who have felt some "negative value" from you as well.  Something to think about.
